README.md
# Catenis API Client for Rust

This library is used to make it easier to access the Catenis API services from code
written in the Rust programming language.

> **Note**: this release of the library targets **version 0.12** of the Catenis API.

## Documentation

The complete library documentation can be found [here](https://docs.rs/catenis_api_client/~3.0).

## Usage

To start using the library, one needs to instantiate a new [`CatenisClient`](https://docs.rs/catenis_api_client/~3.0/catenis_api_client/struct.CatenisClient.html)
object. Then, to make a call to an API method, just call the corresponding method on the client object.

### Example

```rust
use catenis_api_client::{
    CatenisClient, ClientOptions, Environment, Result,
};

// Instantiate Catenis API client object
let mut ctn_client = CatenisClient::new_with_options(
    Some((
        "drc3XdxNtzoucpw9xiRp",
        concat!(
            "4c1749c8e86f65e0a73e5fb19f2aa9e74a716bc22d7956bf3072b4bc3fbfe2a0",
            "d138ad0d4bcfee251e4e5f54d6e92b8fd4eb36958a7aeaeeb51e8d2fcc4552c3"
        ),
    ).into()),
    &[
        ClientOptions::Environment(Environment::Sandbox),
    ],
)?;

// Call Read Message API method
let result = ctn_client.read_message("o3muoTnnD6cXYyarYY38", None)?;

println!("Read message result: {:?}", result);
```

## Notification

The library also makes it easy for receiving notifications from the Catenis system through its
[`WsNotifyChannel`](https://docs.rs/catenis_api_client/~3.0/catenis_api_client/notification/struct.WsNotifyChannel.html)
data structure, which embeds a WebSocket client.

## Asynchronous processing

The library allows for asynchronous processing using the [Tokio](https://crates.io/crates/tokio)
runtime.

To activate asynchronous processing, the **`async`** feature must be enabled.

```toml
catenis_api_client = { version = "3.0", features = ["async"] }
```

The asynchronous version of the client can then be accessed from the [`async_impl`](https://docs.rs/catenis_api_client/~3.0/catenis_api_client/async_impl/index.html)
module.

### Example

```rust
use catenis_api_client::{
    async_impl,
    ClientOptions, Environment, Result,
};

// Instantiate asynchronous Catenis API client object
let mut ctn_client = async_impl::CatenisClient::new_with_options(
    Some((
        "drc3XdxNtzoucpw9xiRp",
        concat!(
            "4c1749c8e86f65e0a73e5fb19f2aa9e74a716bc22d7956bf3072b4bc3fbfe2a0",
            "d138ad0d4bcfee251e4e5f54d6e92b8fd4eb36958a7aeaeeb51e8d2fcc4552c3"
        ),
    ).into()),
    &[
        ClientOptions::Environment(Environment::Sandbox),
    ],
)?;
```

## Catenis API Documentation

For further information on the Catenis API, please reference the [Catenis API Documentation](https://catenis.com/docs/api).

## License

This library is distributed under the terms of both the [MIT License](LICENSE-MIT) and the [Apache License (Version 2.0)](LICENSE-APACHE).

Copyright © 2021-2022, Blockchain of Things Inc.

CHANGELOG.md
# Changelog

## [3.0.1] - 2022-11-14

### Fixes
- Fix the data structures used for issuing request for the API methods Issue Non-Fungible Asset and Reissue Non-Fungible
  Asset.

## [3.0.0] - 2022-11-10

### Breaking changes
- The data structure returned by the *retrieve_asset_info* method of the CatenisClient object has been updated to
  include the new field *is_non_fungible* as per the new behavior of version 0.12 of the Catenis API.
- The data structure returned by the *retrieve_asset_issuance_history* method of the CatenisClient object has been
  updated to account for the two different issuance event types (for regular and for non-fungible assets) that are now
  returned as per the new behavior of version 0.12 of the Catenis API.
- The data structure returned by the *list_permission_events* method of the CatenisClient object has been updated to
  account for the new permission events introduced by version 0.12 of the Catenis API.
- The data structure returned by the *list_notification_events* method of the CatenisClient object and the data
  structure used for processing notification messages (*NotificationMessage*) have been updated to account for the new
  notification events introduced by version 0.12 of the Catenis API.
- Asynchronous processing now requires version 1 of the Tokio runtime.

### New features
- Added support for changes introduced by version 0.12 of the Catenis API: new non-fungible assets feature, including
  the new API methods Issue Non-Fungible Asset, Reissue Non-Fungible Asset, Retrieve Non-Fungible Asset Issuance
  Progress, Retrieve Non-Fungible Token, Retrieve Non-Fungible Token Retrieval Progress, Transfer Non-Fungible Token,
  Retrieve Non-Fungible Token Transfer Progress.

## [2.0.1] - 2021-09-02

### Fixes
- Fix documentation to reference the latest version of the Catenis API client library.

## [2.0.0] - 2021-09-02

### Breaking changes
- The methods *log_message* and *log_chunked_message* of the CatenisClient object have been merged into a single method
  named *log_message*.
- The methods *send_message* and *send_chunked_message* of the CatenisClient object have been merged into a single
  method named *send_message*.
- The data structure returned by the *list_asset_holders* method of the CatenisClient object has been updated to account
  for the new entry that reports the total asset amount that is currently migrated to foreign blockchains as per the new
  behavior of version 0.11 of the Catenis API.

### New features
- Added support for changes introduced by version 0.11 of the Catenis API: new asset export feature, including the new
  API methods Export Asset, Migrate Asset, Asset Export Outcome, Asset Migration Outcome, List Exported Assets, and
  List Asset Migrations.

## [1.0.1] - 2021-01-04

### New features
- Initial version of the library.

src/lib.rs
//! # Catenis API Client
//!
//! This library is used to make it easier to access the Catenis API services.
//!
//! > **Note**: this release of the library targets **version 0.12** of the Catenis API.
//!
//! ## Usage
//!
//! To use it, one needs to instantiate a new [`CatenisClient`] object. Then, to make a call to an
//! API method, just call the corresponding method on the client object.
//!
//! ### Example
//!
//! ```no_run
//! use catenis_api_client::{
//!     CatenisClient, ClientOptions, Environment, Result,
//! };
//!
//! # fn main() -> Result<()> {
//! // Instantiate Catenis API client object
//! let mut ctn_client = CatenisClient::new_with_options(
//!     Some((
//!         "drc3XdxNtzoucpw9xiRp",
//!         concat!(
//!             "4c1749c8e86f65e0a73e5fb19f2aa9e74a716bc22d7956bf3072b4bc3fbfe2a0",
//!             "d138ad0d4bcfee251e4e5f54d6e92b8fd4eb36958a7aeaeeb51e8d2fcc4552c3"
//!         ),
//!     ).into()),
//!     &[
//!         ClientOptions::Environment(Environment::Sandbox),
//!     ],
//! )?;
//!
//! // Call Read Message API method
//! let result = ctn_client.read_message("o3muoTnnD6cXYyarYY38", None)?;
//!
//! println!("Read message result: {:?}", result);
//! # Ok(())
//! # }
//! ```
//!
//! ## Notification
//!
//! The library also makes it easy for receiving notifications from the Catenis system through its
//! [`WsNotifyChannel`] data structure, which embeds a WebSocket client.
//!
//! ## Asynchronous processing
//!
//! The library allows for asynchronous processing using the [Tokio](https://crates.io/crates/tokio)
//! runtime.
//!
//! To activate asynchronous processing, the **`async`** feature must be enabled.
//!
//! ```toml
//! catenis_api_client = { version = "3.0", features = ["async"] }
//! ```
//!
//! The asynchronous version of the client can then be accessed from the [`async_impl`] module.
//!
//! ### Example
//!
//! ```no_run
//! use catenis_api_client::{
//!     async_impl,
//!     ClientOptions, Environment, Result,
//! };
//!
//! # #[tokio::main]
//! # async fn main() -> Result<()> {
//! // Instantiate asynchronous Catenis API client object
//! let mut ctn_client = async_impl::CatenisClient::new_with_options(
//!     Some((
//!         "drc3XdxNtzoucpw9xiRp",
//!         concat!(
//!             "4c1749c8e86f65e0a73e5fb19f2aa9e74a716bc22d7956bf3072b4bc3fbfe2a0",
//!             "d138ad0d4bcfee251e4e5f54d6e92b8fd4eb36958a7aeaeeb51e8d2fcc4552c3"
//!         ),
//!     ).into()),
//!     &[
//!         ClientOptions::Environment(Environment::Sandbox),
//!     ],
//! )?;
//! # Ok(())
//! # }
//! ```

use std::{
    fmt::{
        self,
        Display, Formatter,
    },
};

pub use serde_json::Value as JsonValue;
pub use serde_json::Map as JsonMap;
pub use api::JsonObject;

#[macro_use]
mod macro_impl;

mod base_client;
mod client;
mod date_time;

#[cfg(test)]
mod test_helper;

pub mod api;
pub mod notification;
pub mod error;
#[cfg(feature = "async")]
pub mod async_impl;

use error::GenericError;

pub use client::*;
#[doc(no_inline)]
pub use error::{
    Error, Result,
};
pub use date_time::UtcDateTime;
#[doc(no_inline)]
pub use notification::WsNotifyChannel;

pub(crate) const X_BCOT_TIMESTAMP: &str = "x-bcot-timestamp";
const DEFAULT_BASE_URL: &str = "https://catenis.io/";
const API_BASE_URL_PATH: &str = "api/:version/";
const DEFAULT_API_VERSION: ApiVersion = ApiVersion(0, 12);

type KVList<'a> = &'a [(&'a str, &'a str)];

/// Credentials for a Catenis virtual device to access the Catenis API services.
#[derive(Debug, Clone, Eq, PartialEq)]
pub struct DeviceCredentials {
    /// The device ID.
    pub device_id: String,
    /// Device's API access secret.
    pub api_access_secret: String,
}

impl From<(&str, &str)> for DeviceCredentials {
    fn from(pair: (&str, &str)) -> DeviceCredentials {
        DeviceCredentials {
            device_id: String::from(pair.0),
            api_access_secret: String::from(pair.1),
        }
    }
}

/// Available Catenis API server environments.
#[derive(Debug, Copy, Clone)]
pub enum Environment {
    /// Production environment.
    Prod,
    /// Sandbox environment used for testing.
    Sandbox,
}

/// Specifies a version of the Catenis API.
///
/// # Example
///
/// ```
/// use catenis_api_client::ApiVersion;
///
/// # fn main() {
/// let api_version = ApiVersion(10, 0);
///
/// assert_eq!(api_version.to_string(), "10.0");
/// # }
///```
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct ApiVersion(pub u16, pub u16);

impl Display for ApiVersion {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "{}.{}", self.0, self.1)
    }
}

/// Option settings that can be used when instantiating the Catenis API client.
#[derive(Debug, Copy, Clone)]
pub enum ClientOptions<'a> {
    /// Host name (with optional port) of target Catenis API server. Default value: **`"catenis.io"`**.
    Host(&'a str),
    /// Environment of target Catenis API server. Default value: **`Environment::Prod`**.
    Environment(Environment),
    /// Indicates whether a secure connection (HTTPS) should be used. Default value: **`true`**.
    Secure(bool),
    /// Version of Catenis API to target. Default value: **`ApiVersion(11, 0)`**.
    Version(ApiVersion),
    /// Indicates whether request body should be compressed. Default value: **`true`**.
    UseCompression(bool),
    /// Minimum size, in bytes, of request body for it to be compressed. Default value: **`1024`**.
    CompressThreshold(usize),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_convert_device_credentials_from_pair() {
        let device_credentials = DeviceCredentials::from((
            "drc3XdxNtzoucpw9xiRp",
            "4c1749c8e86f65e0a73e5fb19f2aa9e74a716bc22d7956bf3072b4bc3fbfe2a0d138ad0d4bcfee251e4e5f54d6e92b8fd4eb36958a7aeaeeb51e8d2fcc4552c3"
        ));

        assert_eq!(device_credentials, DeviceCredentials {
            device_id: String::from("drc3XdxNtzoucpw9xiRp"),
            api_access_secret: String::from("4c1749c8e86f65e0a73e5fb19f2aa9e74a716bc22d7956bf3072b4bc3fbfe2a0d138ad0d4bcfee251e4e5f54d6e92b8fd4eb36958a7aeaeeb51e8d2fcc4552c3"),
        });
    }

    #[test]
    fn it_convert_pair_into_device_credentials() {
        let device_credentials: DeviceCredentials = (
            "drc3XdxNtzoucpw9xiRp",
            "4c1749c8e86f65e0a73e5fb19f2aa9e74a716bc22d7956bf3072b4bc3fbfe2a0d138ad0d4bcfee251e4e5f54d6e92b8fd4eb36958a7aeaeeb51e8d2fcc4552c3"
        ).into();

        assert_eq!(device_credentials, DeviceCredentials {
            device_id: String::from("drc3XdxNtzoucpw9xiRp"),
            api_access_secret: String::from("4c1749c8e86f65e0a73e5fb19f2aa9e74a716bc22d7956bf3072b4bc3fbfe2a0d138ad0d4bcfee251e4e5f54d6e92b8fd4eb36958a7aeaeeb51e8d2fcc4552c3"),
        });
    }

    #[test]
    fn it_format_api_version() {
        let api_version = ApiVersion(10, 0);

        assert_eq!(api_version.to_string(), "10.0");
    }
}